### ISSUE TITLE
High losses in performance test

### ISSUE THREAD

**Issue by sergio-gimenez** [p4lang/behavioral-model#823]
Hello!

I am doing a performance test with the bmv2 `simple_switch`. The scenario of the test is Mininet in the [tutorials VM](https://p4.org/events/2019-04-30-p4-developer-day/). The topology consists in two hosts: **h1** and **h11** connected each other through a switch **s1**.

This is the [p4 program](https://github.com/p4lang/behavioral-model/files/3851479/efcp-generator.txt.txt) that basically, regarding this test, parse ethernet and IPv4 headers, lookup to the table and then forwards the packet.

I also have done a c program ([ip-generator.c](https://github.com/p4lang/behavioral-model/files/3851483/ip-generator.c.txt)) that sends packets from h1 (sender) to h11(recevier). These are the results of the test:

![Throughput  Mbps  vs Packet Size  bytes](https://user-images.githubusercontent.com/42917235/68943758-5f212b00-07ac-11ea-926b-5b5e81131cf3.png)

All the tests have been done without log macros (`--disable-logging-macros` flag).

Shouldn't I expect a higer throughput for such a simple topology/resources use? But that's not all, I am asking myself if the great amount of lost packets in the router interfaces is the expected.

```
s1-eth1   Link encap:Ethernet  HWaddr 00:00:00:00:01:0b  
          inet6 addr: fe80::200:ff:fe00:10b/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:3 errors:0 dropped:0 overruns:0 frame:0
          TX packets:3512050 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:270 (270.0 B)  TX bytes:2062470204 (2.0 GB)

s1-eth2   Link encap:Ethernet  HWaddr 00:00:00:00:01:01  
          inet6 addr: fe80::200:ff:fe00:101/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:3535557 errors:0 dropped:0 overruns:0 frame:0
          TX packets:7 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:2082070512 (2.0 GB)  TX bytes:750 (750.0 B)
```

The snippet above shows the bmv2 interfaces directly connected to the hosts(`s1-eth1` receiver and `s1-eth2` sender) during all the test. I'm quite surprised with the amount of lost packets:  
Received packets in `s1-eth2` `RX packets:3535557` and forwarded packets in `s1-eth1` `TX packets:3512050` entails losses around **90%**.

I changed the inter arrival time of the packets in order to evaluate the losses:

![Losses vs  inter arrival time  μs](https://user-images.githubusercontent.com/42917235/68943783-719b6480-07ac-11ea-8dbc-91a44e9343ad.png)

So, all in all:

* Which throughput should I expect? I was told in another issue that it was about 100 kpps, which is quite lower than what I have now.

* Is it normal to have such a big amount of losses?

* How can this results can be improved? Any suggestions?

**Comment by jafingerhut** [comment]
I have some experiment results of my own detailed below, if you are curious.

But first, why do you care to improve the results?  simple_switch is intended for development and debug, not high speed packet forwarding.  It was never advertised as high performance.  What development/debug scenarios lead you to want more?

I tried with a different, also small, P4 program that has only 3 table lookups per packet, with only 1 table entry installed in each of the 3 tables, and sending the same identical packet that matched that 1 entry in every packet.  I sent packets using iperf at 100Mbps with 512-byte UDP packets (so send rate about 24K packets/sec), and received and dropped packets also using iperf.  I used tcpdump writing to a file on the sending and receiving virtual Ethernet interfaces, and while all the packets were sent across the sending interface before going to simple_switch, only about 20% of them were received, so about 20Mbps received.

I ran 'top' while the experiment was going on, and the only processes that used much CPU time were simple_switch, which went up to around 189% CPU (so 2 CPU cores almost 100% utilized, in parallel), and iperf, which was under 50% of another core, on a 4 vCPU virtual machine.

I compiled simple_switch with debug options enabled on the 'configure' command by using these commands: https://github.com/jafingerhut/p4-guide/blob/master/bin/install-p4dev-v2.sh#L291-L318 .  I started simple_switch with these command line options:
```
sudo simple_switch -i 0@veth2 -i 1@veth4 -i 2@veth6 -i 3@veth8 -i 4@veth10 -i 5@veth12 -i 6@veth14 -i 7@veth16 demo1.p4_16.json
```
Where the BMv2 JSON file was compiled from this source code: https://github.com/jafingerhut/p4-guide/blob/master/demo1/demo1.p4_16.p4

If I used the default compiler options to build simple_switch by running `./configure` instead, I suspect the throughput would be somewhat better, but probably not more than 2x or so.  I can try if you are curious.

That is around 4000 to 4500 packets/sec forwarded, and the rest dropped.

**Comment by jafingerhut** [comment]
I repeated my experiment after building simple_switch with the default "./configure", so I believe invokes the C compiler with -O2 optimization level, instead of whatever it is with the configure command I used before.  It got up to about 7,800 packets/sec received, so nearly double what I reported before.  The data rate was about 32 Mbits/sec, but I believe it is the packets/sec that is more the limiting factor, with close to the same packets/sec regardless of their size.

**Comment by antoninbas** [comment]
The issue you opened in the tutorials repo showed a much higher throughput in the 100s of Mbps: https://github.com/p4lang/tutorials/issues/303. That's more in line with what I would expect. What happened since then?

**Comment by antoninbas** [comment]
I'm getting 250Mbps on a larger program (can be reproduced by running `sudo python stress_test_ipv4.py` inside the mininet/ directory) with 3 tables (2 exact match, one LPM).

However CPU usage is very low: 75% so less than one core. That means there is something else at play here. I have observed this is in the past when running the benchmark inside a local VM, like I'm doing now. I'll try on an AWS EC2 instance when I get the chance. If you have a Linux server available, you may want to try it natively. It should be around 1Gbps with that program.

Another thing: I am using an old JSON file produced by the legacy p4c-bm compiler from a P4_14 program. These JSON files tend to be more efficient, with fewer arithmetic operations and copies.

**Comment by antoninbas** [comment]
And I used `./configure 'CXXFLAGS=-g -O3' 'CFLAGS=-g -O3' --disable-logging-macros --disable-elogger`

**Comment by antoninbas** [comment]
I tried on a reasonable AWS instance (`c4.2xlarge`) which has 8 vCPUs and got the following results:
```
Ready !
*** Ping: testing ping reachability
h1 -> h2
h2 -> h1
*** Results: 0% dropped (2/2 received)
Running iperf measurement 1 of 5
1056 Mbps
Running iperf measurement 2 of 5
1056 Mbps
Running iperf measurement 3 of 5
1047 Mbps
Running iperf measurement 4 of 5
1042 Mbps
Running iperf measurement 5 of 5
1041 Mbps
Median throughput is 1047 Mbps
```

CPU usage was slightly over 400% for simple_switch and around 50% for iperf (sender + receiver). That CPU usage, along with the 1+Gbps throughput (80+kpps),  matches what I would expect.

I think there are some I/O issues when running the benchmark inside a local VM. Surprisingly, while I observe this issue for Linux VMs running on a MacOS host, I never had this issue for Linux VMs running on a Linux host (the ec2 instance is obviously also a VM and does not have that I/O issue). If anyone has an explanation, I am very curious to hear it...

Regarding packet losses. Nothing surprising here. If simple_switch cannot keep-up, packets are going to be dropped at the sender's interface.

**Comment by jafingerhut** [comment]
So not too surprisingly, the performance one gets depends on a lot of factors, such as:

+ which P4 program you are running
+ what P4 compiler you used to compile it to a BMv2 JSON file with, since the later ones in some cases produce less efficient BMv2 JSON expressions that take longer to execute during packet processing
+ which version of behavioral-model code you are running, and what optimization settings you configured it with for the C++ compiler to compiler behavioral-model's code
+ what command line options you give to simple_switch when starting it
+ how many table entries you install in the tables being applied while processing packets
+ how many CPU cores you have on the system, and how many the traffic pattern you send to simple_switch enables it to take advantage of, e.g. I believe if you send input packets that are all from one input port, and all destined to one output port, simple_switch can use at most one CPU core for ingress, plus one more for egress, or something like that?  But if you send packets into different input ports, going to different output ports, in some cases more CPU cores can be utilized, because simple_switch can run different ingress threads for different input ports, and similarly different egress threads for packets going to different output ports?
+ whether all packets are hitting the same table entries as each other, or some large mix of different table entries, which would affect how many cache hits vs. cache misses the CPU would get while processing packets.
+ whether you are running on a bare metal Linux machine, or a Linux VM, and even whether that Linux VM is running on a macOS host OS, or a Linux host OS.
+ probably 17 other factors I'm not even imagining right now :-)

This is where the phrase YMMV comes from.

**Comment by jafingerhut** [comment]
And when I used the options Antonin mentioned to compile behavioral-model:
```
./configure 'CXXFLAGS=-g -O3' 'CFLAGS=-g -O3' --disable-logging-macros --disable-elogger
```
and by the way all of my experimental results reported in this issue were in a Linux VM running on a macOS host, for 512-byte UDP packets I was getting much better performance with 200% CPU utilization for simple_switch (2 CPU cores utilized 100% each) of about 75,000 packets/sec, which is about 300 Mbits/sec.  Not too shabby.

**Comment by jafingerhut** [comment]
Another detail: even though my test could achieve up to 75,000 packets/sec forwarded by simple_switch, there were a noticeable percentage of packet losses even when I was sending much lower packet rates.  I suspect there is some kind of burstiness either in how they are generated and enqueued, or in how they are dequeued by the simple_switch process, combined with what must be a limited buffer size in the operating system's kernel for each veth interface, that is smaller than required to completely avoid packet losses at lower average packet rates, even though simple_switch can keep up with the average rate just fine.

**Comment by sergio-gimenez** [comment]
First of all, thanks a lot to both of you @antoninbas and @jafingerhut for the fast and complete answers you have give to my question. Your help, not only in this issue, but also in many others, has been very useful for my work. It's not much, but you will be acknowleged in the paper I'm publishing in a few weeks as a thankful gesture :grin: :grin:.

Regarding the performance test. As @antoninbas proposed, I tried it on an AWS EC2 instance (`c4.2xlarge`). Following step by step the guide in the [performance document](https://github.com/p4lang/behavioral-model/blob/master/docs/performance.md), i.e running the [simple_router.p4](https://github.com/p4lang/behavioral-model/blob/master/mininet/simple_router.p4) program compiled with the old-fastish compiler, I got an average throughput around **630 Mbps**. I also tried it with my own [p4 program](https://github.com/p4lang/behavioral-model/files/3868852/efcpRouter.p4.txt) compiled with the [p4c docker image](https://hub.docker.com/r/p4lang/p4c) and I have obtained an average throughput around **590 Mbps**.

I went through the python [stress_test_ipv4](https://github.com/p4lang/behavioral-model/blob/master/mininet/stress_test_ipv4.py.in) source file and I changed the iperf command `return h1.popen("iperf -f m -c %s -t 30" % (h2.IP()), stdout=subprocess.PIPE)` to `return h1.popen("iperf -f m -c %s -u -b 2G -l 1472" % (h2.IP()), stdout=subprocess.PIPE)
` in the client side and   `server = h2.popen("iperf -s -u")` to ` server = h2.popen("iperf -s")` in the server side in order to send UDP packets instead of TCP, which I think it makes more sense for a performance test. Now with my p4 program I'm getting this results:

![UDP Throughput  Mbps  vs Packet Size  bytes](https://user-images.githubusercontent.com/42917235/69233915-21e2e180-0b8e-11ea-9ccd-4feb5c73ab53.png)

Also I don't really understand  why I have this drop in the throughput value at 250... :sweat_smile:

So, all in all, thank you guys for all the help given :).

**Comment by jafingerhut** [comment]
Except for the packet sizes under 250 bytes or so, which I cannot explain either, off the top of my head, that graph is probably easiest to interpret if you change the Y axis to be "Throughput (packets/sec)", since then it appears that for 250 bytes and up, it would be a constant packet rate, or close to it.
